File: app/src/main/java/com/itus/u_money/view/fragment/TransactionFragment.java
package com.itus.u_money.view.fragment;

import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.github.mikephil.charting.animation.Easing;
import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;
import com.github.mikephil.charting.formatter.PercentFormatter;
import com.github.mikephil.charting.utils.ColorTemplate;
import com.github.mikephil.charting.utils.MPPointF;
import com.itus.u_money.R;
import com.itus.u_money.databinding.FragmentTransactionBinding;
import com.itus.u_money.contract.TransactionContract;
import com.itus.u_money.model.Icon;
import com.itus.u_money.model.Transaction;
import com.itus.u_money.model.TransactionType;
import com.itus.u_money.presenter.TransactionPresenter;
import com.itus.u_money.view.adapter.SummaryAdapter;
import com.itus.u_money.view.adapter.TransactionAdapter;
import com.itus.u_money.view.utils.SpacingItemDecoration;

import java.util.ArrayList;
import java.util.List;

public class TransactionFragment extends Fragment implements TransactionContract.View {
   private static TransactionFragment mInstance = null;
   private PieChart chart;
   FragmentTransactionBinding binding;
   private TransactionContract.Presenter presenter;
   private TransactionAdapter transactionAdapter;
   private SummaryAdapter summaryAdapter;
   public static TransactionFragment getInstance() {
      if (mInstance == null)
         mInstance = new TransactionFragment();
      return mInstance;
   }
   public TransactionFragment() {
      presenter = new TransactionPresenter(this);
   }

   @Nullable
   @Override
   public View onCreateView(
           @NonNull LayoutInflater inflater,
           @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
      binding = DataBindingUtil.inflate(inflater, R.layout.fragment_transaction, container, false);
      binding.setPresenter(presenter);
      return binding.getRoot();
   }

   @Override
   public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
      super.onViewCreated(view, savedInstanceState);
      initChart();
      initRecyclerView(binding.recyclerviewListTransaction);
      presenter.onViewCreated();
   }
   @Override
   public void onResume() {
      super.onResume();
      presenter.getListTransaction();
   }

   private void initRecyclerView(RecyclerView recyclerView) {
      transactionAdapter = new TransactionAdapter();
      recyclerView.setLayoutManager(new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false));
      recyclerView.setAdapter(transactionAdapter);
      recyclerView.addItemDecoration(new SpacingItemDecoration(12));
   }

   void initChart() {
      chart = getActivity().findViewById(R.id.chart);
      chart.setUsePercentValues(true);
      chart.getDescription()
           .setEnabled(false);
      chart.setExtraOffsets(5, 10, 5, 5);

      chart.setDragDecelerationFrictionCoef(0.95f);
      chart.setDrawHoleEnabled(false);
      chart.setHoleColor(Color.WHITE);

      chart.setTransparentCircleColor(Color.WHITE);
      chart.setTransparentCircleAlpha(110);

      chart.setRotationAngle(0);
      // enable rotation of the chart by touch
      chart.setRotationEnabled(true);
      chart.setHighlightPerTapEnabled(true);
      // chart.spin(2000, 0, 360);

      chart.getLegend()
           .setEnabled(false);

      // entry label styling
      chart.setEntryLabelColor(Color.WHITE);
      chart.setEntryLabelTextSize(12f);

      summaryAdapter = new SummaryAdapter();
      binding.recyclerviewSummary.setLayoutManager(new LinearLayoutManager(getContext(), RecyclerView.HORIZONTAL, false));
      binding.recyclerviewSummary.setAdapter(summaryAdapter);
   }

   @Override
   public void updateReportBy(String text, boolean increase) {
      binding.textReportBy.setText(text);
      Animation animation;
      if (increase)
         animation = AnimationUtils.loadAnimation(getContext(), R.anim.up_in);
      else
         animation = AnimationUtils.loadAnimation(getContext(), R.anim.down_in);
      animation.reset();
      binding.textReportBy.clearAnimation();
      binding.textReportBy.startAnimation(animation);
   }

   @Override
   public void updateDate(String text) {
      binding.tvDate.setText(text);
   }

   @Override
   public void upDateChart(boolean next) {
      Animation animationOut, animationIn;
      if (next) {
         animationOut = AnimationUtils.loadAnimation(getContext(), R.anim.right_out);
         animationIn = AnimationUtils.loadAnimation(getContext(), R.anim.left_in);
      } else {
         animationOut = AnimationUtils.loadAnimation(getContext(), R.anim.left_out);
         animationIn = AnimationUtils.loadAnimation(getContext(), R.anim.right_in);
      }
      animationOut.setAnimationListener(new Animation.AnimationListener() {
         @Override
         public void onAnimationStart(Animation animation) {

         }
         @Override
         public void onAnimationEnd(Animation animation) {
            chart.clearAnimation();
            chart.startAnimation(animationIn);
         }
         @Override
         public void onAnimationRepeat(Animation animation) {

         }
      });
      chart.startAnimation(animationOut);
   }
   @Override
   public void animateChart() {
      chart.animateY(500, Easing.EaseInOutQuad);
   }

   @Override
   public void showTransactionList(List<Transaction> transactionList, List<TransactionType> transactionTypes, List<Icon> icons) {
      transactionAdapter.setTransactionList(transactionList);
      transactionAdapter.setTransactionTypeList(transactionTypes);
      transactionAdapter.setIconList(icons);
      transactionAdapter.notifyDataSetChanged();
   }

   @Override
   public void updateChartData(List<TransactionType> transactionTypes, List<Icon> icons, boolean animate) {
      binding.recyclerviewListTransaction.removeAllViews();
      summaryAdapter.setData(transactionTypes);
      summaryAdapter.notifyDataSetChanged();

      ArrayList<PieEntry> entries = new ArrayList<>();
      for (int i = 0; i < transactionTypes.size(); i++) {
         TransactionType transactionType = transactionTypes.get(i);
         entries.add(new PieEntry(transactionType.total, transactionType.name));
      }

      PieDataSet dataSet = new PieDataSet(entries, "Election Results");
      dataSet.setDrawIcons(false);
      dataSet.setSliceSpace(3f);
      dataSet.setIconsOffset(new MPPointF(0, 40));
      dataSet.setSelectionShift(5f);
      ArrayList<Integer> colors = new ArrayList<>();
      for (Icon icon : icons)
         colors.add(Color.parseColor(icon.backgroundColor));

      dataSet.setColors(colors);
      PieData data = new PieData(dataSet);
      data.setValueFormatter(new PercentFormatter(chart));
      data.setValueTextSize(11f);
      data.setValueTextColor(Color.WHITE);
      chart.setData(data);
      if (animate)
         chart.animateY(500, Easing.EaseInOutQuad);
      chart.highlightValues(null);
      chart.invalidate();
   }
}

File: app/src/main/java/com/itus/u_money/view/fragment/TransactionFragment.kt
package com.itus.u_money.view.fragment

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.github.mikephil.charting.animation.Easing
import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry
import com.github.mikephil.charting.formatter.PercentFormatter
import com.github.mikephil.charting.utils.MPPointF
import com.itus.u_money.R
import com.itus.u_money.contract.DetailTransactionContract
import com.itus.u_money.contract.TransactionContract
import com.itus.u_money.databinding.FragmentTransactionBinding
import com.itus.u_money.model.Icon
import com.itus.u_money.model.Transaction
import com.itus.u_money.model.TransactionType
import com.itus.u_money.presenter.TransactionPresenter
import com.itus.u_money.view.activity.DetailTransactionActivity
import com.itus.u_money.view.adapter.SummaryAdapter
import com.itus.u_money.view.adapter.TransactionAdapter
import com.itus.u_money.view.adapter.TransactionListener
import com.itus.u_money.view.utils.SpacingItemDecoration
import java.util.*

class TransactionFragment : Fragment(), TransactionContract.View, TransactionListener {
    private var chart: PieChart? = null
    var binding: FragmentTransactionBinding? = null
    private val presenter: TransactionContract.Presenter
    private var transactionAdapter: TransactionAdapter? = null
    private var summaryAdapter: SummaryAdapter? = null
    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_transaction, container, false)
        binding!!.presenter = presenter
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initChart()
        initRecyclerView(binding!!.recyclerviewListTransaction)
        presenter.onViewCreated()
    }

    override fun onResume() {
        super.onResume()
        presenter.getListTransaction()
    }

    private fun initRecyclerView(recyclerView: RecyclerView) {
        transactionAdapter = TransactionAdapter(this)
        recyclerView.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        recyclerView.adapter = transactionAdapter
        recyclerView.addItemDecoration(SpacingItemDecoration(12))
    }

    private fun initChart() {
        val chart: PieChart = activity!!.findViewById(R.id.chart)
        with(chart) {
            setUsePercentValues(true)
            description.isEnabled = false
            setExtraOffsets(5f, 10f, 5f, 5f)
            dragDecelerationFrictionCoef = 0.95f
            isDrawHoleEnabled = false
            setHoleColor(Color.WHITE)
            setTransparentCircleColor(Color.WHITE)
            setTransparentCircleAlpha(110)
            rotationAngle = 0f
            // enable rotation of the chart by touch
            isRotationEnabled = true
            isHighlightPerTapEnabled = true
            // chart.spin(2000, 0, 360);
            legend.isEnabled = false
            // entry label styling
            setEntryLabelColor(Color.WHITE)
            setEntryLabelTextSize(12f)
        }

        summaryAdapter = SummaryAdapter()
        binding!!.recyclerviewSummary.layoutManager = LinearLayoutManager(context, RecyclerView.HORIZONTAL, false)
        binding!!.recyclerviewSummary.adapter = summaryAdapter

        this.chart = chart
    }

    override fun updateReportBy(text: String, increase: Boolean) {
        binding!!.textReportBy.text = text
        val animation: Animation = if (increase) AnimationUtils.loadAnimation(context, R.anim.up_in) else AnimationUtils.loadAnimation(context, R.anim.down_in)
        animation.reset()
        binding!!.textReportBy.clearAnimation()
        binding!!.textReportBy.startAnimation(animation)
    }

    override fun updateDate(text: String) {
        binding!!.tvDate.text = text
    }

    override fun upDateChart(next: Boolean) {
        val animationOut: Animation
        val animationIn: Animation
        if (next) {
            animationOut = AnimationUtils.loadAnimation(context, R.anim.right_out)
            animationIn = AnimationUtils.loadAnimation(context, R.anim.left_in)
        } else {
            animationOut = AnimationUtils.loadAnimation(context, R.anim.left_out)
            animationIn = AnimationUtils.loadAnimation(context, R.anim.right_in)
        }
        animationOut.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {}
            override fun onAnimationEnd(animation: Animation) {
                chart!!.clearAnimation()
                chart!!.startAnimation(animationIn)
            }

            override fun onAnimationRepeat(animation: Animation) {}
        })
        chart!!.startAnimation(animationOut)
    }

    override fun animateChart() {
        chart!!.animateY(500, Easing.EaseInOutQuad)
    }

    override fun showTransactionList(transactionList: List<Transaction>, transactionTypes: List<TransactionType>, icons: List<Icon>) {
        transactionAdapter!!.setTransactionList(transactionList)
        transactionAdapter!!.setTransactionTypeList(transactionTypes)
        transactionAdapter!!.setIconList(icons)
        transactionAdapter!!.notifyDataSetChanged()
    }

    override fun updateChartData(transactionTypes: List<TransactionType>, icons: List<Icon>, animate: Boolean) {
        binding!!.recyclerviewListTransaction.removeAllViews()
        summaryAdapter!!.setData(transactionTypes)
        summaryAdapter!!.notifyDataSetChanged()
        val entries = ArrayList<PieEntry>()
        for (i in transactionTypes.indices) {
            val transactionType = transactionTypes[i]
            val name: String? = transactionType.name
            entries.add(PieEntry(transactionType.total.toFloat(), name))
        }
        val dataSet = PieDataSet(entries, "Election Results")
        dataSet.setDrawIcons(false)
        dataSet.sliceSpace = 3f
        dataSet.iconsOffset = MPPointF(0f, 40f)
        dataSet.selectionShift = 5f
        val colors = ArrayList<Int>()
        for (icon in icons) colors.add(Color.parseColor(icon.backgroundColor))
        dataSet.colors = colors
        val data = PieData(dataSet)
        data.setValueFormatter(PercentFormatter(chart))
        data.setValueTextSize(11f)
        data.setValueTextColor(Color.WHITE)
        chart!!.data = data
        if (animate) chart!!.animateY(500, Easing.EaseInOutQuad)
        chart!!.highlightValues(null)
        chart!!.invalidate()
    }

    override fun onTransactionClick(transaction: Transaction) {
        val intent = Intent(context, DetailTransactionActivity::class.java)
        intent.putExtra("Transaction", transaction)
        startActivity(intent)
    }

    companion object {
        private var mInstance: TransactionFragment? = null

        @JvmStatic
        val instance: TransactionFragment?
            get() {
                if (mInstance == null) mInstance = TransactionFragment()
                return mInstance
            }
    }

    init {
        presenter = TransactionPresenter(this)
    }
}